File: 백준/Silver/24445. 알고리즘 수업 － 너비 우선 탐색 2/알고리즘 수업 － 너비 우선 탐색 2.py
import sys
from collections import deque
input = sys.stdin.readline

def bfs(x):
    global cnt
    q = deque([x])
    visited[x] = cnt
    while q:
        now = q.popleft()
        for i in graph[now]:
            if not visited[i]:
                q.append(i)
                cnt += 1
                visited[i] = cnt

n, m, r = map(int, input().split())
graph = [[] for _ in range(n+1)]
visited = [0]*(n+1)
cnt = 1
for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
for g in graph:
    g.sort(reverse=True)

bfs(r)
print(*visited[1:], sep='\n')

File: 백준/Silver/24445. 알고리즘 수업 － 너비 우선 탐색 2/알고리즘 수업 － 너비 우선 탐색 2.java
import java.io.IOException;
import java.util.*;

public class Main {
    private static List<List<Integer>> graph;
    private static int[] visited;
    private static int cnt;
    private static StringBuilder sb;

    public static void main(String[] args) throws IOException {
        sb = new StringBuilder();
        int n = readInt(), m = readInt(), r = readInt();
        visited = new int[n + 1];
        graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            int u = readInt(), v = readInt();
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        graph.forEach(node -> node.sort(Collections.reverseOrder()));

        bfs(r);

        for (int i = 1; i <= n; i++) {
            sb.append(visited[i]).append('\n');
        }

        System.out.println(sb);
    }

    private static void bfs(int start) {
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(start);
        visited[start] = ++cnt;

        while (!queue.isEmpty()) {
            int x = queue.poll();
            for (int i = 0; i < graph.get(x).size(); i++) {
                int y = graph.get(x).get(i);
                if (visited[y] == 0) {
                    queue.offer(y);
                    visited[y] = ++cnt;
                }
            }
        }
    }

    private static int readInt() throws IOException {
        int t = 0;
        byte v;
        while ((v = read()) <= 32);
        do t = (t << 3) + (t << 1) + (v & 15);
        while (isNumber(v = read()));
        return t;
    }

    static final int SIZE = 1 << 13;
    static byte[] buffer = new byte[SIZE];
    static int index, size;
    private static byte read() throws IOException {
        if (index == size) {
            size = System.in.read(buffer, index = 0, SIZE);
            if (size < 0) buffer[0] = -1;
        }
        return buffer[index++];
    }

    static boolean isNumber(byte c) {
        return 47 < c && c < 58;
    }
}